wgsl/index.bs

```diff
@@ -12172,6 +12263,48 @@ executing a [=statement/discard=] statement), execution of the quad may be
 terminated; however, such termination is not considered to produce [=uniform
 control flow|non-uniform control flow=].
 
+## Subgroups ## {#subgroups}
+
+A <dfn noexport>subgroup</dfn> is a set of invocations which concurrently
+execute a [=compute shader stage|compute=] or [=fragment shader
+stage|fragment=] shader stage [=entry point=] that can efficiently share data
+and [[#subgroup-ops|collectively]] compute results.
+In a compute shader, each subgroup is a subset of a particular workgroup.
+In a fragment shader, a subgroup might contain invocations from multiple [=draw commands=].
+Each [=quad=] will be contained in a single subgroup.
+
+The <dfn noexport>subgroup size</dfn> is the maximum number of invocations in a subgroup.
+This value is accessible via the [=built-in values/subgroup_size=] built-in value.
+Subgroup size is a [=uniform value=] within a [=compute shader stage/workgroup=],
+but not within a [=draw command=].
+All subgroup sizes are within the range [4, 128].
+The actual size is a function of the device properties and device compiler.
+Each device supports a subset of the possible range of subgroup sizes (possibly a single value).
+The device compiler selects a size from the supported sizes using a variety of heuristics.
+Each subgroup may contain fewer invocations than the reported subgroup size
+(e.g. if fewer invocations than the subgroup size are launched).
+
+An invocation's <dfn noexport>subgroup invocation ID</dfn> a unique id within the subgroup.
+This id is accessible via the [=built-in values/subgroup_invocation_id=] built-in value.
+There is no defined relationship between subgroup_invocation_id and [=built-in values/local_invocation_index=].
+Users should not assume a particular mapping in order to avoid non-portable code.
+
+Due to non-uniform control flow, subgroups may diverge during execution.
+An invocation is <dfn noexport dfn-for="subgroups">active</dfn> if it is executing a particular operation.
+Invocations that are not executing code due to control flow are considered
+<dfn noexport dfn-for="subgroups">inactive</dfn>.
+When the subgroup size is greater than the number of invocation in a subgroup, the extra invocations are
+considered inactive.
+[=Helper invocations=] may be active or inactive in an operation.
+That is, on some devices helper invocations may participate in subgroup operations,
+while on other devices they may not.
+
+Note: There is considerable non-portability among underlying devices when
+operating in non-uniform control flow and device compilers often aggressively
+optimize such code.
+The result is that the subgroup may contain a different set of active
+invocations than the shader author expects.
```

Comment:
Maybe link in to the new uniformity analysis again, telling authors that there is more portability if they abide by them?

Reply:
It is noted below, but IDK that shader authors would go read `### Subgroup Operations ### {#subgroup-ops}` itself instead of the intro + definition of builtins.

Reply:
The surrounding text was rewritten, please take another look.